File: 2277-count-equal-and-divisible-pairs-in-an-array/count-equal-and-divisible-pairs-in-an-array.java
class Solution {
    public int countPairs(int[] nums, int k) {
        int count = 0 ;
        HashMap<Integer,List<Integer>> mp = new HashMap<>();
        for(int i=0 ; i<nums.length;i++){
            if(mp.containsKey(nums[i])){
                List<Integer> curr = new LinkedList<>(mp.get(nums[i]));
                    for(int j=0;j<curr.size();j++){
                        if(curr.get(j)*i % k == 0) count++;
                    }
                    mp.get(nums[i]).add(i);
                }
                else{
                    List<Integer> curr = new LinkedList<>();
                    curr.add(i);
                    mp.put(nums[i],curr);
                }
            }
            return count;
        }
    }

File: 2277-count-equal-and-divisible-pairs-in-an-array/count-equal-and-divisible-pairs-in-an-array.cpp
class Solution {
public:
    int countPairs(vector<int>& nums, int k) {
        int count = 0;
        unordered_map<int, vector<int>> mp;

        for (int i = 0; i < nums.size(); i++) {
            if (mp.find(nums[i]) != mp.end()) {
                vector<int>& curr = mp[nums[i]];
                for (int j = 0; j < curr.size(); j++) {
                    if ((curr[j] * i) % k == 0) {
                        count++;
                    }
                }
                mp[nums[i]].push_back(i);
            } else {
                mp[nums[i]] = {i};
            }
        }

        return count;
    }
};